File: programmers/2022_KAKAO/04/one.py
def getSubset():
    res = []

    def dfs(start=10, arr=[]):
        res.append(arr)

        for i in range(start,-1,-1):
            dfs(i - 1, arr+[i])
    
    dfs()
    return res

def isPosible(c, info, n):
    sum = 0
    for i in c:
        sum += info[i]+1
    return True if sum <= n else False

def calcScore(c, info):
    s = 0
    for i in range(11):
        if i in c:
            s += 10-i
        elif info[i] != 0:
            s -= 10-i
    return s

def solution(n, info):
    answer = []
    candidates = getSubset()
    max_val = -1
    for c in candidates:
        if isPosible(c, info, n):
            s = calcScore(c, info)
            if s > max_val:
                max_val = s
                answer = c
    
    if len(answer) == 0 or max_val == 0:
        return [-1]
    res = [info[s]+1 if s in answer else 0 for s in range(11)]
    if sum(res)< n:
        res[-1] += n-sum(res)
    return res

print(getSubset())

File: programmers/2022_KAKAO/04/one.js
const getSubset = () => {
  let res = [];
  const dfs = (start = 9, arr = []) => {
    res.push(arr);
    for (let i = start; i > -1; i--) {
      dfs(i - 1, [...arr, i]);
    }
  };

  dfs();
  return res;
};

function solution(n, info) {
  var answer = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
  const cadidates = getSubset();
  //   const cadidates = [[0, 1]];
  let maxDiff = Number.MIN_SAFE_INTEGER;
  for (let wins of cadidates) {
    // 1. 화살 수 가능한지 확인
    const arrows = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    let a = 0;
    for (let w of wins) {
      a += info[w] + 1; // 10-w 점 획득
      arrows[w] = info[w] + 1;
    }
    if (a > n) {
      continue;
    }
    // 2. 점수차 계산
    let score = 0;
    for (let i = 0; i < 11; i++) {
      if (wins.includes(i)) {
        score += 10 - i;
      } else if (info[i] !== 0) {
        score -= 10 - i;
      }
    }
    arrows[10] = n - arrows.reduce((a, b) => a + b, 0);
    if (0 < score && maxDiff < score) {
      maxDiff = score;
      answer = arrows;
    }
  }
  if (maxDiff === Number.MIN_SAFE_INTEGER) {
    answer = [-1];
  }
  return answer;
}

n = 5;
info = [2, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0];

console.log(solution(n, info));
// console.log(3 in [1, 2, 3]);
